frontend/controller/app/identity.js

```diff
@@ -0,0 +1,376 @@
+'use strict'
+
+import { GIErrorUIRuntimeError, L, LError, LTags } from '@common/common.js'
+import sbp from '@sbp/sbp'
+import { LOGIN, LOGIN_COMPLETE, LOGIN_ERROR, LOGOUT } from '~/frontend/utils/events.js'
+import { Secret } from '~/shared/domains/chelonia/Secret.js'
+import { boxKeyPair, buildRegisterSaltRequest, computeCAndHc, decryptContractSalt, hash, hashPassword, randomNonce } from '~/shared/zkpp.js'
+// Using relative path to crypto.js instead of ~-path to workaround some esbuild bug
+import * as Common from '@common/common.js'
+import { cloneDeep, has } from '@model/contracts/shared/giLodash.js'
+import { CURVE25519XSALSA20POLY1305, EDWARDS25519SHA512BATCH, deriveKeyFromPassword, serializeKey } from '../../../shared/domains/chelonia/crypto.js'
+import { handleFetchResult } from '../utils/misc.js'
+
+const { Vue } = Common
+
+const loadState = async (identityContractID: string, password: ?string) => {
+  if (password) {
+    const stateKeyEncryptionKeyFn = (stateEncryptionKeyId, salt) => {
+      return deriveKeyFromPassword(CURVE25519XSALSA20POLY1305, password, salt + stateEncryptionKeyId)
+    }
+
+    const { encryptionParams, value: state } = await sbp('gi.db/settings/loadEncrypted', identityContractID, stateKeyEncryptionKeyFn)
+
+    if (state) {
+      // If state contains a saved Chelonia state, extract it.
+      // The saved Chelonia state will be used to restore Chelonia on this
+      // new session.
+      const cheloniaState = state.cheloniaState
+      delete state.cheloniaState
+
+      return { encryptionParams, state, cheloniaState }
+    } else {
+      // There's no state to restore
+      return { encryptionParams, state, cheloniaState: null }
+    }
+  } else {
+    const state = await sbp('gi.db/settings/load', identityContractID)
+
+    // cheloniaState is only stored in settings when logging in with a password
+    // If there's an active session, then chelnoiaState is stored and
+    // managed separately
+    return { encryptionParams: null, state, cheloniaState: null }
+  }
+}
+
+sbp('okTurtles.events/on', LOGIN, async ({ identityContractID, encryptionParams, state }) => {
+  // This function restores the state after a successful login. It's complex
+  // because it deals with _three_ different states.
+  // * `state` (input parameter): (optional). This is the _saved_ Vuex state
+  //   that we'll use to replace the current Vuex state (`vuexState`).
```

Comment:
Could you further add commentary here on when `state` is passed in and when it isn't?